manager/allocator/network/ipam/ipam.go

```diff
@@ -0,0 +1,729 @@
+			Network:              local.nw,
+			Aliases:              config.Aliases,
+			DriverAttachmentOpts: config.DriverAttachmentOpts,
+			Addresses:            make([]string, 0, len(config.Addresses)),
+		}
+		// even before we've allocated anything, add the attachment to the
+		// finalAttachments, so that the roll back proceeds correctly if
+		// anything fails
+		finalAttachments = append(finalAttachments, attachment)
+
+		// now get the IPAM driver and options
+		ipam, _ := a.drvRegistry.IPAM(local.nw.IPAM.Driver.Name)
+		ipamOpts := local.nw.IPAM.Driver.Options
+
+		// now allocate all of the addreses. The user might not have requested
+		// any addreses specifically, in which case we should just add one
+		// emptystring for the loop body to run once
+		addresses := config.Addresses
+		if len(addresses) == 0 {
+			addresses = []string{""}
+		}
+	addressesLoop:
+		for _, address := range addresses {
+			var requestIP net.IP
+			if address != "" {
+				var err error
+				requestIP, _, err = net.ParseCIDR(address)
+				if err != nil {
+					requestIP = net.ParseIP(address)
+					if requestIP == nil {
+						return nil, ErrInvalidAddress{address}
+					}
+				}
+			}
+			// now that we've got the IP address in a usable form, try
+			// reserving it. we will need to try for each pool
+			for _, poolID := range local.pools {
+				ip, _, err := ipam.RequestAddress(poolID, requestIP, ipamOpts)
+				if err == nil {
+					// we only want 1 address per task. so, once we've found a
+					// valid address, break out of the addresses loop, and go
+					// to the next attachmnet
+					local.endpoints[ip.String()] = poolID
+					attachment.Addresses = append(attachment.Addresses, ip.String())
+					break addressesLoop
+				}
+				// if we get ErrIPOutOfRange or ErrNoAvailableIPs, it means
+				// this pool might not be right or available for this address.
+				// in that case, try the next pool
+				if err == ipamapi.ErrIPOutOfRange || err == ipamapi.ErrNoAvailableIPs {
+					continue
+				}
+				// if we get any other error, bail out because this address
+				// isn't going to work
+				return nil, ErrFailedAddressRequest{address, err}
+			}
+		}
+	}
+	// finally, once we've gone through every address for every config, return
+	// the final result and no error
+	return finalAttachments, nil
+}
+
+func (a *allocator) DeallocateAttachments(attachments []*api.NetworkAttachment) {
+	a.deallocateAttachments(attachments)
+}
+
+func (a *allocator) deallocateAttachments(attachments []*api.NetworkAttachment) {
+	for _, attachment := range attachments {
+		// get the local network and IPAM driver
+		local := a.networks[attachment.Network.ID]
+		ipam, _ := a.drvRegistry.IPAM(local.nw.IPAM.Driver.Name)
+
+		for _, address := range attachment.Addresses {
+			poolID := local.endpoints[address]
+			delete(local.endpoints, address)
+
+			// both of these things can error, but we can't do anything about
+			// it if they do, so we ignore the errors.
```

Comment:
Should we log a debug log if they error?  Also when we clean up networks, we return a `ErrDoubleFault` if the cleanup fails.  Should we also do that when cleaning up attachments when rolling back?

Reply:
Again, the error handling for this kind of sucks. Error on deallocation should *never* happen, because if they do, there's something wrong, and we don't really have an opportunity for a do-over or a way to sanely handle them. I think that I will change this, like above, to return errors to the caller, and let the caller decide what to do, which will eventually involve this getting kicked up to the top level and logged there.